enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,435 @@
+The proposed Node Healthcheck Controller (NHC) provides a generic solution that fits all installation types, together with an external remediation API that allows a pluggable remediation system to be installed and remediate unhealthy nodes.
+
+### Goals
+
+- Create an installation-type agnostic node health check solution
+
+- Feature parity with Machine Health Check controller
+
+- Providing a default strategy combined with [poison-pill].
+
+- Endorse usage of external [remediation api]
+
+- Allow defining different unhealthy criteria by node selectors, and have a sane strategy for handling nodes that match multiple NHC CRs.
+
+- Provide defaults health checks and default remedy custom resource creation.
+
+- Long term - Make Machine Health Check controller only implement the external remediation api and drop the default Machine API based remediation mechanism
+
+
+### Non-Goals
+
+- Coordination of remedy operation - operators/controllers implementers
+expected to track the [node maintenance lease proposal]
+
+- Recording long-term stable history of all health-check failures or remediation
+  actions.
+
+- Detection of Machines that never become healthy Nodes is not part of the initial implementation.  Machine API based cluster can/should continue to use MHC at this time.
+
+## Proposal
+The main use-case we target is UPI, user provisioned infrastructure, which do not have
+the Machine-API controlling them. This enables any kind of recovery mechanism to be built and handle remediation.
+
+#### Stories
+- As an admin of a cluster I want the cluster to detect Node failures and
+ initiate recovery, so that workloads and volumes with at-most-one semantics
+  can be run/mounted elsewhere.
+
+- As an admin of a cluster I want the cluster to detect Node failures and
+ initiate recovery, so that we can attempt to restore cluster capacity.
+
+- As an OCP engineer, I want the NHC to consume the same logic for detecting
+ failures, to avoid code duplication and provide a consistent experience
+  compared to MHC.
+
+- As an OCP engineer, I want the NHC to use the same interface for initiating
+ recovery, so that we can reuse existing mechanisms (like [poison pill] and metal3) and
+  avoid code duplication.
+
+To fulfil the above we create a [Node Health Check custom resource (NCR)](anchor todo) that specifies:
+1. health criteria: list of criterias and selector
+2. template: a template of an [external custom resource (remediation CR)](link todo)
+ to create in response to a match.
+
+### Unhealthy criteria
+A node target is unhealthy when the node meets the unhealthy node conditions criteria defined.
+If any of those criterias are met for longer than given timeouts, the controller
+creates a remediation CR according to the template in the NHC CRD.
+Timeouts will be defined as part of the NHC CRD .
+
+### Implementation Details
```

Comment:
What is the plan to package and deliver this new component? Is it going into the OCP release payload, or are we going to deliver it through OLM?

Reply:
OLM. See suggest revision to the summary